core/services/headtracker/head_listener.go

```diff
@@ -0,0 +1,258 @@
+		Name: "head_listener_heads_received",
+		Help: "The total number of heads seen",
+	})
+	promEthConnectionErrors = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "head_listener_eth_connection_errors",
+		Help: "The total number of eth node connection errors",
+	})
+)
+
+type Config interface {
+	EthHeadTrackerMaxBufferSize() uint
+	EthereumURL() string
+	ChainID() *big.Int
+}
+
+type HeadListener struct {
+	config           Config
+	ethClient        eth.Client
+	headers          chan *models.Head
+	headSubscription ethereum.Subscription
+	connectedMutex   sync.RWMutex
+	connected        bool
+	sleeper          utils.Sleeper
+
+	log      *logger.Logger
+	muLogger sync.RWMutex
+
+	chStop chan struct{}
+	wgDone *sync.WaitGroup
+}
+
+func NewHeadListener(l *logger.Logger,
+	ethClient eth.Client,
+	config Config,
+	chStop chan struct{},
+	wgDone *sync.WaitGroup,
+	sleepers ...utils.Sleeper,
+) *HeadListener {
+	var sleeper utils.Sleeper
+	if len(sleepers) > 0 {
+		sleeper = sleepers[0]
+	} else {
+		sleeper = utils.NewBackoffSleeper()
+	}
+	return &HeadListener{
+		config:    config,
+		ethClient: ethClient,
+		sleeper:   sleeper,
+		log:       l,
+		chStop:    chStop,
+		wgDone:    wgDone,
+	}
+}
+
+// SetLogger sets and reconfigures the log for the head tracker service
+func (hl *HeadListener) SetLogger(logger *logger.Logger) {
+	hl.muLogger.Lock()
+	defer hl.muLogger.Unlock()
+	hl.log = logger
+}
+
+func (hl *HeadListener) logger() *logger.Logger {
+	hl.muLogger.RLock()
+	defer hl.muLogger.RUnlock()
+	return hl.log
+}
+
+func (hl *HeadListener) ListenForNewHeads(handleNewHead func(ctx context.Context, header models.Head) error, connected func()) {
+	defer hl.wgDone.Done()
+	defer func() {
+		if err := hl.unsubscribeFromHead(); err != nil {
+			hl.logger().Warn(errors.Wrap(err, "HeadListener failed when unsubscribe from head"))
+		}
+	}()
+
+	ctx, cancel := utils.ContextFromChan(hl.chStop)
+	defer cancel()
+
+	for {
+		if !hl.subscribe(connected) {
+			break
+		}
+		err := hl.receiveHeaders(ctx, handleNewHead)
+		if ctx.Err() != nil {
+			break
+		} else if err != nil {
+			hl.logger().Errorw(fmt.Sprintf("Error in new head subscription, unsubscribed: %s", err.Error()), "err", err)
+			hl.headers = nil
+			continue
+		} else {
+			break
+		}
+	}
+}
+
+// This should be safe to run concurrently across multiple nodes connected to the same database
+// Note: returning nil from receiveHeaders will cause listenForNewHeads to exit completely
+func (hl *HeadListener) receiveHeaders(ctx context.Context, handleNewHead func(ctx context.Context, header models.Head) error) error {
+	for {
+		select {
+		case <-hl.chStop:
+			return nil
+		case blockHeader, open := <-hl.headers:
+			if !open {
+				return errors.New("HeadTracker: headers prematurely closed")
+			}
+			if blockHeader == nil {
+				hl.logger().Error("HeadTracker: got nil block header")
+				continue
+			}
+			promNumHeadsReceived.Inc()
+
+			err := handleNewHead(ctx, *blockHeader)
```

Comment:
Where is this defined? I'd expect to find it in the same file?

Perhaps we need a `hl.handleNewHead` instead

Reply:
oh i had it as a method parameter:
```
func (hl *HeadListener) receiveHeaders(ctx context.Context, handleNewHead func(ctx context.Context, header models.Head) error) error {
```
but yeah not very usual - will make it a field